### ISSUE TITLE
support the use of slot in layout files

### ISSUE THREAD

**Issue by dohomi** [nuxt/nuxt.js#190]
Hello,

it would be nice if `slot` areas would assignable in layout files. Often parts of the pages need to be rendered inside of the top toolbar. Thanks

<div align="right"><sub><em>This question is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c162">#c162</a>)</em></sub></div>

**Comment by Atinux** [comment]
Can you provide an example of what you are trying to do @dohomi ?

**Comment by dohomi** [comment]
Hello @Atinux here a simple example to get content into the toolbar and content section of an app

index.vue
```
<template>
  <div>
    <div slot="toolbar">
      <h2 class="md-title">Welcome to index</h2>
    </div>
    <div slot="content">
      <p>her comes the content...</p>
    </div>
 </div>
</template>
```
defaul_layout.vue
```
<template>
<div>
<header>
  <slot name="toolbar"></slot>
</header>
<content>
  <slot name="content"></slot>
</content>
</div>
</template>
```

Currently I do this through a custom component and leave the `layout` file with just the `<nuxt/>` almost empty.

**Comment by Atinux** [comment]
Hi @dohomi 

It's not possible with `layouts` logic of nuxt.js since it's based on `vue-router`.

But nothing is impossible with vue.js 😄 

You can create a `layout` folder (without the `s`) for example where you can create your layouts with slots.

`layout/default.vue`
```vue
<template>
  <div>
    <header>
      <slot name="toolbar"></slot>
    </header>
    <content>
      <slot name="content"></slot>
    </content>
  </div>
</template>
```

And then, in your pages, you can can import the layout (we use it as `<page>`) and use the slots:
```vue
<template>
  <page>
    <div slot="toolbar">
      <h2 class="md-title">Welcome to index</h2>
    </div>
    <div slot="content">
      <p>her comes the content...</p>
    </div>
 </page>
</template>

<script>
import page from '~/layout/default'

export default {
  components: { page }
}
</script>
```

**Comment by dohomi** [comment]
Hi @Atinux thats how I am doing it right now. Thanks for the example and good to know that I did it in the right way.

**Comment by dohomi** [comment]
Hi @Atinux @alexchopin is it somehow possible to use `router-view` feature of `vue-router`? In my current setup I would love to render the toolbar directly from page into a separate section of the layout. Similiar approach like above, but currently with using a `component` for layout the render of specific fields of my layout are rendered again. Thanks for heads up

**Comment by dohomi** [comment]
never mind... I got it working with tweaking the css and now the structure respects. Still there might be some workflows where `router-view` support would be beneficial for the developers.

**Comment by Etheryte** [comment]
`vue-router` supports named views out of the box, it's very unclear why Nuxt is limiting the usage as it is. Named views and/or slots offer a much more diverse approach to composition and there's no reason to support them only through workarounds.  
[From `vue-router` docs](https://router.vuejs.org/en/essentials/named-views.html):
```html
<router-view class="view one"></router-view>
<router-view class="view two" name="a"></router-view>
<router-view class="view three" name="b"></router-view>
```
```js
const router = new VueRouter({
  routes: [
    {
      path: '/',
      components: {
        default: Foo,
        a: Bar,
        b: Baz
      }
    }
  ]
})
```

**Comment by easherma** [comment]
Any updates on this with current/forthcoming versions?

**Comment by Atinux** [comment]
We will work on it, after long thoughts I don’t think layouts was a great idea but instead rely on a App.vue

**Comment by redijedi** [comment]
Any update on this? I've found the lack of slots really hard to work around. I often need to change the contents of a default toolbar in the layout depending on certain pages' actions needs. I've been bending over backwards to try to handle this. Vue slots were so easy to use and would be great to have again.

**Comment by xxRockOnxx** [comment]
I need this one too. Aside from what @Atinux suggested, is it possible to programatically or suggested to programatically inject a component to layout's component?

Same use case as the others which is modifying a toolbar's contents.

**Comment by xxRockOnxx** [comment]
I'm not sure if this is possible, but what if nuxt kind of do something like this behind the scene?

```
<template>
    <default-layout>
        // actual contents of your page/*.vue
    </default-layout>
</template>
```

so that the slots in the layout is accessible within page components?

**Comment by dylanmcgowan** [comment]
To complete the trifecta of cartoon network avatars +1 for layout slots!

**Comment by geocine** [comment]
I hope this issue could be reopened

**Comment by sobolevn** [comment]
Very sad that this is not supported.

**Comment by nelsonlarocca** [comment]
Hi guys !

Any update about this ?
Thanks !

**Comment by cain** [comment]
Would love to see this!

**Comment by ngunyimacharia** [comment]
Any update on this? Would love to see it implemented!

**Comment by connecteev** [comment]
+1 can we get this feature, please? +1 for layout slots. I hope someone is reading this, otherwise I'll open a new ticket and reference this.

PS. Not sure if Nuxt has changed significantly from 2017 when this comment was made:
https://github.com/nuxt/nuxt.js/issues/1079

**Comment by iliyaZelenko** [comment]
[Portal-vue](https://github.com/LinusBorg/portal-vue) helps me.

**layouts/default.vue**
```vue
<template>
  <div>
    My header

    <PortalTarget name="under-header" />

    <nuxt />
  </div>
</template>
```


**pages/index.vue**
```vue
<template>
  <div>
    <Portal to="under-header">
      My under header
    </Portal>

    My body
  </div>
</template>
```

It works for me.

**Comment by douira** [comment]
That seems like a possibility but not a very nice one structurally... But for small changes, it's just fine. You might not want to use too many of those portals though, it might get confusing otherwise.
My solution is to use a base layout that I extend with other layouts. Then I can specify a different layout when I need it to be slightly different. This approach doesn't work if you need to tightly integrate the slot content with the rest of the page though. (and you can't or don't want to use vuex)

**Comment by connecteev** [comment]
@douira @iliyaZelenko  interesting. Curious how would you recommend creating "filtered" pages similar to this?

- https://dev.to/
- https://dev.to/top/week
- https://dev.to/top/month

and similarly
https://dev.to/t/beginners
https://dev.to/t/beginners/top/week

Using layouts? child pages? We obviously want to be DRY (dont-repeat-yourself) so code reuse is the main goal here.

**Comment by douira** [comment]
Those pages seem to have the same layout so I don't really know what you're getting at. Do you mean how the URLs change what tab is opened? Does that even have anything to do with nuxt layouts? Wouldn't one just use the same layout for all of those and then detect what route is active? Or talk to the router to find out what route is active.

**Comment by connecteev** [comment]
@douira yes...how the URLs change what tab is opened..."week" "month" etc. 
Would you create pages like pages/ {week, month, year} and under what subfolder structure? I'm asking how to best structure the code, and if you have a good github repo of detecting the active route would help...forgive the n00b question.

**Comment by douira** [comment]
That's off-topic for this issue. Contact me somewhere else or open an issue on the nuxt cmty for that.

**Comment by connecteev** [comment]
@douira ok i cleaned up the comment history, lmk how to reach you if you're open to it

**Comment by dhritzkiv** [comment]
portal-vue looks promising, but the lack of SSR support is a non-starter for me.

FWIW, it appears that Vue.js v3 will have portals as a first-class feature: https://github.com/vuejs/vue-next

**Comment by connecteev** [comment]
+1 for slots in layout files so we can get SSR support

**Comment by douira** [comment]
@ the stale bot, this is a feature request so there isn't much to say about that.

**Comment by connecteev** [comment]
@stale bot: please keep this open...pretty please?

**Comment by xfyre** [comment]
Would love to have this feature and have layouts in SSR directly supported.

**Comment by manutheblacker** [comment]
How can i achieve this ?

**Comment by chadwtaylor** [comment]
+1

**Comment by prikr** [comment]
+100!

**Comment by pmanikas** [comment]
Would be really helpful.

**Comment by BeWe11** [comment]
Portal-vue worked for me to inject nav drawer content per page, though I don't need SSR for this content. Would be nice to have support for slots in layouts without this hack... +1

**Comment by rifton007** [comment]
+1

**Comment by rifton007** [comment]
Thanks, but in my case I prefer to not use a additional package.

Finally, I use component like a layout, here a example :

**components/grid/default.vue**
``` html
<template>
  <nav>
    <slot name="nav"/>
  </nav>
  <main>
     <slot/> //default slot
  </main>
  <footer>
    <slot name="footer"/>
  </footer>
  
</template>
```

**pages/index.vue**
``` html
<template>
  <GridDefault> // component
    <nav slot="nav">...</nav>
    <h1>Some page</h1>
    <p>Some paragraph</p>
    <div slot="footer">...</div>
  </GridDefault>
</template>
```

**Comment by 19peaches** [comment]
Yeah using a component as a root element doesn't behave as expected.

I've tried this as in the first example from way back, creating a `page` component and then using that is a wrapping root element.

The wrapping root component doesn't get access to root instance data props, or even vuex states.

I have a `page.props` data property set on my root instance, when I look at it in vue dev tools, it's completely populated with data, but yet the when I try to access it `page.props.user` in the template it throws an error.

I also have this same data stored in vuex state, but when attempting to access it I get null.

```javascript
import { mapGetters } from 'vuex'

export default {
    mounted() {
        console.log(this.pageProps) /// logs out null
    },
    computed: {
        ...mapGetters({
            pageProps: 'base/page'
        }),
    }
} 
```

I suppose it might suffice if you only want to put slots in it, but it won't render any data.

Layouts need to render slots, just like Vue does.

**Comment by dawadam** [comment]
@mrleblanc101 how is this possible in Nuxt 3 ?
